File: tools/jlpt-import/schema.py
"""
Nguồn sự thật duy nhất cho hình dạng JSON đề JLPT.

Khớp 1:1 với mô hình dữ liệu ở docs/jlpt-practice-test-research.md mục 10 và định dạng
nhập ở mục 11.3. Mọi script trong tools/jlpt-import/ phải build dữ liệu qua các hàm ở đây,
không tự ghép dict tay ở nơi khác — để không có hai script tình cờ sinh ra hai hình dạng
JSON khác nhau.
"""

from __future__ import annotations

from dataclasses import dataclass, field, asdict
from typing import Literal, Optional

JlptLevel = Literal["N5", "N4", "N3", "N2", "N1"]

ScoringSection = Literal["gengo_chishiki", "dokkai", "choukai"]

MondaiType = Literal[
    # 文字・語彙
    "kanji_yomi", "hyouki", "bunmyaku_kitei", "iikae_ruigi", "youhou",
    # 文法
    "bunpou_keishiki", "bun_no_kumitate", "bunshou_no_bunpou",
    # 読解
    "naiyou_tan", "naiyou_chuu", "naiyou_chou", "jouhou_kensaku",
    # 聴解
    "kadai_rikai", "point_rikai", "gaiyou_rikai", "hatsuwa_hyougen", "sokuji_outou",
]

# Nhãn tiếng Nhật + tiếng Anh chuẩn cho từng khối, dùng khi build TimedBlock.
BLOCK_LABELS: dict[str, tuple[str, str]] = {
    "moji_goi": ("言語知識（文字・語彙）", "Vocab"),
    "bunpou_dokkai": ("言語知識（文法）・読解", "Grammar & Reading"),
    "choukai": ("聴解", "Listening"),
}


@dataclass
class JlptChoice:
    text: str
    note: Optional[str] = None
    linkedItemKey: Optional[str] = None


@dataclass
class JlptQuestion:
    id: str
    level: JlptLevel
    mondai: MondaiType
    scoringSection: ScoringSection
    choices: list[JlptChoice]
    answerIndex: int
    stem: Optional[str] = None
    stemUnderline: Optional[tuple[int, int]] = None
    explanation: Optional[str] = None
    passageId: Optional[str] = None
    audioId: Optional[str] = None
    transcript: Optional[str] = None


@dataclass
class MondaiGroup:
    mondai: MondaiType
    instruction: str
    questionIds: list[str]


@dataclass
class TimedBlock:
    id: str
    label: str
    minutes: int
    mondai: list[MondaiType]
    labelEn: Optional[str] = None


@dataclass
class JlptExam:
    id: str
    level: JlptLevel
    title: str
    blocks: list[TimedBlock]
    groups: list[MondaiGroup]
    questionIds: list[str]
    source: Literal["original", "official-sample", "user-provided"] = "user-provided"


@dataclass
class JlptImportFile:
    """Hình dạng đúng như mô tả ở mục 11.3 — đây là cái ghi ra output/*.json."""

    formatVersion: int
    exam: JlptExam
    groups: list[MondaiGroup]
    questions: list[JlptQuestion]

    def to_dict(self) -> dict:
        return asdict(self)


def missing_choice_notes(questions: list[JlptQuestion]) -> list[str]:
    """Id các câu còn thiếu lời giải cho ít nhất một phương án (mục 8.7, 11.6).

    Dùng để in cảnh báo sau khi bóc — PDF đề thi gốc không có sẵn lời giải,
    nên gần như chắc chắn TOÀN BỘ câu vừa bóc sẽ nằm trong danh sách này.
    """
    out = []
    for q in questions:
        if any(not c.note for c in q.choices):
            out.append(q.id)
    return out

File: src/lib/jlpt/schema.ts
/**
 * Mô hình dữ liệu đề JLPT — khớp mục 10 và định dạng nhập ở mục 11.3 của
 * docs/jlpt-practice-test-research.md. Đừng sửa các type này mà không đọc lại tài liệu đó,
 * vì file nhập do AI khác soạn ở ngoài repo cũng phải khớp đúng hình dạng này.
 */

export type JlptLevel = 'N5' | 'N4' | 'N3' | 'N2' | 'N1';

export type ScoringSection = 'gengo_chishiki' | 'dokkai' | 'choukai';

export const MONDAI_TYPES = [
  'kanji_yomi', 'hyouki', 'bunmyaku_kitei', 'iikae_ruigi', 'youhou',
  'bunpou_keishiki', 'bun_no_kumitate', 'bunshou_no_bunpou',
  'naiyou_tan', 'naiyou_chuu', 'naiyou_chou', 'jouhou_kensaku',
  'kadai_rikai', 'point_rikai', 'gaiyou_rikai', 'hatsuwa_hyougen', 'sokuji_outou',
] as const;
export type MondaiType = (typeof MONDAI_TYPES)[number];

export interface TimedBlock {
  id: string;
  label: string;
  labelEn?: string;
  minutes: number;
  mondai: MondaiType[];
}

export interface MondaiGroup {
  mondai: MondaiType;
  instruction: string;
  questionIds: string[];
}

export interface JlptExam {
  id: string;
  level: JlptLevel;
  title: string;
  blocks: TimedBlock[];
  groups: MondaiGroup[];
  questionIds: string[];
  source: 'original' | 'official-sample' | 'user-provided';
}

export interface JlptChoice {
  text: string;
  note?: string;
  linkedItemKey?: string;
}

export interface JlptQuestion {
  id: string;
  level: JlptLevel;
  mondai: MondaiType;
  scoringSection: ScoringSection;
  stem?: string;
  stemUnderline?: [number, number];
  choices: JlptChoice[];
  answerIndex: number;
  explanation?: string;
  furigana?: { text: string; reading: string }[];
  passageId?: string;
  audioId?: string;
  transcript?: string;
  transcriptAnswerSpan?: [number, number];
  grammarPoint?: string;
  confusableWith?: string[];
  vocabIds?: string[];
  kanjiChars?: string[];
}

/** 読解: nhiều câu cùng trỏ về một đoạn văn (mục 7.3 và mục 10). */
export interface Passage {
  id: string;
  level: JlptLevel;
  kind: 'tan' | 'chuu' | 'chou' | 'jouhou';
  text: string;
  source?: string;
}

/** Hình dạng của một file nhập (dán JSON / tải file), theo mục 11.3. */
export interface JlptImportFile {
  formatVersion: 1;
  exam: {
    id: string;
    level: JlptLevel;
    title: string;
    source?: 'original' | 'official-sample' | 'user-provided';
    blocks: TimedBlock[];
  };
  groups: MondaiGroup[];
  questions: JlptQuestion[];
  /** Tuỳ chọn — chỉ cần khi có câu 読解 dùng chung đoạn văn. */
  passages?: Passage[];
}

/** Một đề đã lưu — gói cả file nhập gốc lẫn thông tin quản lý để hiện trong danh sách. */
export interface StoredJlptExam {
  exam: JlptExam;
  questions: JlptQuestion[];
  passages: Passage[];
  /** Đề do AI sinh chưa được người kiểm lại thì đánh dấu, không tính vào thống kê tiến bộ (mục 11.9). */
  reviewed: boolean;
  importedAt: number;
  updatedAt: number;
}

// ─── Lượt làm bài (mục 10) ───────────────────────────────────────────

export type Confidence = 'sure' | 'unsure' | 'guess';

export interface JlptAnswer {
  questionId: string;
  chosenIndex: number | null;
  confidence: Confidence;
  flagged: boolean;
  timeSpentMs: number;
  changeCount: number;
}

export type AttemptStatus = 'running' | 'paused' | 'submitted' | 'reviewing' | 'reviewed' | 'abandoned';
export type AttemptMode = 'taste' | 'section' | 'full';

export interface JlptAttempt {
  id: string;
  examId: string;
  level: JlptLevel;
  status: AttemptStatus;
  mode: AttemptMode;
  /** Câu hỏi thuộc phiên này, theo đúng thứ tự làm bài (phụ thuộc mode). */
  questionIds: string[];
  startedAt: number;
  submittedAt?: number;
  /**
   * Hạn nộp bài (epoch ms), tính lúc `createAttempt()` từ tổng phút của (các) khối tính giờ
   * liên quan — xem `attemptLogic.ts`. `undefined` = không có áp lực thời gian gắt (mode
   * `taste`, mục 5.1: phiên "nhấm nháp" cố ý không đếm ngược). Lưu thẳng vào attempt (không
   * tính lại mỗi lần mở màn) để F5/đóng tab quay lại vẫn tính đúng giờ còn lại, giống cách
   * `ExamSession.tsx` lưu `deadline` vào phiên đang làm dở của mình.
   */
  deadline?: number;
  answers: Record<string, JlptAnswer>;
  predictedPercent?: number;
  reviewedQuestionIds: string[];
  /**
   * Chủ sở hữu lượt làm bài: id tài khoản, hoặc null/thiếu = làm ở chế độ khách.
   *
   * Đề là kho chung nhưng LƯỢT LÀM BÀI là của riêng từng người, kể cả khi nhiều người
   * dùng chung một trình duyệt — xem src/lib/jlpt/db.ts.
   */
  ownerId?: string | null;
  /** % đúng lúc nộp bài, chốt sẵn để trang chủ khỏi phải nạp lại cả đề để tính điểm. */
  scorePercent?: number;
  /**
   * Các câu làm sai, chốt sẵn lúc nộp bài.
   *
   * Cùng lý do với `scorePercent`: biết "còn bao nhiêu câu chưa mổ xẻ" (hiệu số với
   * `reviewedQuestionIds`) mà không phải nạp lại cả đề — trang chủ và danh sách đề cần con
   * số này cho mọi đề cùng lúc. Lượt làm bài từ trước khi có trường này sẽ thiếu, khi đó
   * phải tính lại bằng `scoreAttempt()` — xem `wrongIdsOf()` trong attemptLogic.ts.
   */
  wrongQuestionIds?: string[];
}

// ─── Sổ tay lỗi riêng cho JLPT (mục 6.2-6.4) ─────────────────────────

export type MistakeCause = 'goi' | 'bunpou' | 'kanji' | 'dokkai' | 'choukai' | 'wana' | 'bat_can' | 'het_gio';

export const MISTAKE_CAUSES: { code: MistakeCause; label: string; hint: string }[] = [
  { code: 'goi', label: 'Không biết từ', hint: 'Thiếu từ vựng' },
  { code: 'bunpou', label: 'Không nắm ngữ pháp', hint: 'Chưa biết, hoặc lẫn hai mẫu gần nghĩa' },
  { code: 'kanji', label: 'Sai chữ Hán', hint: 'Đọc sai âm, nhầm chữ giống nhau' },
  { code: 'dokkai', label: 'Hiểu sai đoạn văn', hint: 'Đọc lướt, bỏ sót từ nối, hiểu ngược ý' },
  { code: 'choukai', label: 'Nghe sót / nghe nhầm', hint: 'Không kịp, nhầm âm gần giống' },
  { code: 'wana', label: 'Dính bẫy đề', hint: 'Đáp án "trông có vẻ đúng"' },
  { code: 'bat_can', label: 'Bất cẩn', hint: 'Biết mà chọn nhầm' },
  { code: 'het_gio', label: 'Không kịp giờ', hint: 'Chưa kịp đọc đã phải đoán' },
];

export interface MistakeEntry {
  id: string;
  /** Chủ sở hữu — cùng quy ước với JlptAttempt.ownerId. */
  ownerId?: string | null;
  questionId: string;
  examId: string;
  attemptId: string;
  createdAt: number;
  cause: MistakeCause;
  confidenceAtAnswer: Confidence;
  chosenIndex: number | null;
  reattemptIndex?: number | null;
  myRule?: string;
  myExample?: string;
  /** Khoá thẻ SRS liên quan, nếu câu này (hoặc đáp án đúng) nối được với thẻ đã có. */
  srsKey?: string;
}
